### DISCUSSION TITLE
_Literal_zero_is_expected() referenced in <compare> not defined anywhere?

### DISCUSSION THREAD

**Discussion by DTBob** [microsoft/STL#4697] [General] [answered]
I am getting a link error with an unresolved external void std::_Literal_zero_is_expected() used in function std::_Literal_zero. I am no expert in the stl, so I ask this question here rather than assuming it is an issue - is/should _Literal_zero_is_expected be actually defined anywhere? I have cloned and searched the stl and not found it. Cheers.

**Comment by StephanTLavavej** [comment, marked as answer]
It's intentionally never defined. If you're seeing this, you've got code somewhere that's attempting to compare a spaceship operator result against an int that isn't a literal `0`.

```
C:\Temp>type meow.cpp
```
```cpp
#include <compare>
#include <print>
#include <string>
using namespace std;

int main() {
    const string cats{"cats"};
    const string dogs{"dogs"};
    println("(cats <=> dogs) < 0: {}", (cats <=> dogs) < 0);
#ifndef SKIP_BAD_CODE
    println("(cats <=> dogs) < 5: {}", (cats <=> dogs) < 5);
#endif
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od /DSKIP_BAD_CODE meow.cpp && meow
meow.cpp
(cats <=> dogs) < 0: true

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /MTd /Od meow.cpp && meow
meow.cpp
meow.cpp(11): error C7595: 'std::_Literal_zero::_Literal_zero': call to immediate function is not a constant expression
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.41.33901\include\compare(31): note: failure was caused by call of undefined function or one not declared 'constexpr'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.41.33901\include\compare(31): note: see usage of 'std::_Literal_zero_is_expected'
meow.cpp(11): note: the call stack of the evaluation (the oldest call first) is
meow.cpp(11): note: while evaluating function 'std::_Literal_zero::_Literal_zero<int,0>(_Ty) noexcept'
        with
        [
            _Ty=int
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.41.33901\include\compare(31): note: while evaluating function 'void std::_Literal_zero_is_expected(void)'
```

**Reply by DTBob** [reply]
OK, thanks for that explanation Stephan.  So, how does this work when statically linking - if _Literal_zero_is_expected() is purposefully not defined, then static linking fails, so how do we statically link?

**Comment by cpplearner** [comment]
A compiler should never generate a call to `_Literal_zero_is_expected` in the binary, because `_Literal_zero_is_expected` is only used in a `consteval` constructor, which should be evaluated at compile time.

It's possible that you encountered a compiler bug, but a reproducible example is needed for further analysis.
